# Reference Sources

## _Trichomonas vaginalis_ G3 aliases

The gene aliases for _T. vaginalis_ G3 strain were obtained from TrichDB (https://trichdb.org/trichdb/app). The file can found here: https://trichdb.org/trichdb/app/downloads/Current_Release/TvaginalisG3/txt/TrichDB-59_TvaginalisG3_GeneAliases.txt

The table of gene aliases was downloaded from TrichDB on 27.04.2022, with the following protocol: 
 1. Open TrichDB database: https://trichdb.org/trichdb/app/
 2. On the sidebar, click on "Organisms", then in the drop-down menu that appears select "Organisms: Genome Info and Stats"
 3. Select "_Trichomonas vaginalis_ G3"
 4. Scroll down the page to select "Download Genome Data from TrichDB"
 5. Click on "txt/" (the following information is also provided: "[last modified: ]2022-04-13 11:15   [Size:] -   [Description] Codon usage or gene alias tables in tab delimited text file.")
 6. Select "TrichDB-57_TvaginalisG3_GeneAliases.txt" (the following information is also provided: "[Last modified:] 2022-04-13 11:15   [Size:] 3.3M   [Description] Gene information table")
 7. The webpage provides no inbuilt way to save the text file displayed. I did the following: CTRL+A → CTRL+C → CRTL+V in a local text file on my computer named "TrichDB-57_TvaginalisG3_GeneAliases.txt"


## ALE Supplement Files

### Species Tree

The species tree (SpeciesTree_rooted_at_outgroup_22_NameEdit.nwk) is derived from one of the versions of a species tree generated by OrthoFinder (SpeciesTree_rooted_at_outgroup_22.txt). 

The version of the Newick tree utilized as input for R and ALE has been edited in a few ways: 
 - The species names have been made to match the ones used in the data table used as input for Count
 - The species names need to be in quotes (this is required by the Count program)
 - Fixed the naming of _Tritrichomonas foetus_ (file name incorrectly uses _Trichomonas foetus_)


### ALE Nodes Indexing File

When ALE predicts the likelihood of DTL events at each node, it uses its own node numbering system. However, when a tree is imported into R, R uses its own internal numbering system - it cannot accept external numbering systems. The ALE_2R_Indexing.txt file can therefore be used as an optional input to the parse_ALE_Nodes__v2.py and parse_ALE_Nodes_GFam.py programs to output the data table with the proper node labeling needed for R to annotate the tree. 

This indexing file was manually generated by comparing the ALE and R node numbering systems. 
